Add Rust client tutorial + samples, redo transactions in all languages (#5780)
edgedb/edgedb: docs/clients/rust/queryable.rst

.. _ref_rust_queryable:

Using the ``Queryable`` macro
-----------------------------

The easiest way to unpack an EdgeDB query result is the built-in
``Queryable`` macro from the ``edgedb-derive`` crate. This turns queries
directly into Rust types without having to match on a ``Value`` (more in
the section on the ``Value`` enum), cast to JSON, etc.

.. code-block:: rust

  #[derive(Debug, Deserialize, Queryable)]
  pub struct QueryableAccount {
      pub username: String,
      pub id: Uuid,
  }

  let query = "select account {
      username,
      id
      };";
  let as_queryable_account: QueryableAccount = client
      .query_required_single(query, &())
      .await?;

> Note: Field order within the shape of the query matters when using the
> ``Queryable`` macro. In the example below, a query is done in the order
> ``id, username`` instead of ``username, id`` as defined in the struct:

Reviewer: ```suggestion
.. note::

    Field order within the shape of the query matters when using the
    ``Queryable`` macro. In the example below, we run a query with the order
    ``id, username`` instead of ``username, id`` as defined in the struct:
```

.. code-block:: rust

  let query = "select account {
      id,
      username
      };";
  let wrong_order: Result<QueryableAccount, _> = client
      .query_required_single(query, &())
      .await;
  assert!(
      format!("{wrong_order:?}")
      .contains(r#"WrongField { unexpected: "id", expected: "username" }"#);
  );

You can use `cargo expand`_ with the nightly compiler to see the code
generated by the ``Queryable`` macro, but the minimal example repo also

> contains a somewhat cleaned up version of the generated Queryable code:

Reviewer: ```suggestion
contains a somewhat cleaned up version of the generated ``Queryable`` code:
```
